### ISSUE TITLE
may not want Xlib

### ISSUE THREAD

**Issue by yinwang0** [cisco/ChezScheme#9]
After configure and make, I'm getting this error complaining that I don't have Xlib on my mac. I needed to install XQuartz for it to build.

Chez Scheme is a compiler. It may be confusing that the build includes graphics stuff by default. Many people may not want to have X11 on the mac, so maybe it's good to have an option to skip the build of X11 related things?

Or even better, the configure script can check whether the user have X11 and skip the build for that part if they don't have it.

```
gcc -m64 -Wpointer-arith -Wall -Wextra -O2 -I/opt/X11/include/ -c -DX86_64 -I../boot/a6osx -I../zlib expeditor.c
expeditor.c:846:10: fatal error: 'X11/Xlib.h' file not found
#include <X11/Xlib.h>
         ^
1 error generated.
```

**Comment by johnwcowan** [comment]
It's not just a compiler, it's also a run-time library including SWL and the optional SWL-based IDE.

**Comment by yinwang0** [comment]
I know it contains a GUI library and an IDE, but from a user's perspective, they only care about the compiler. Most people would just want to use it in the terminal or with Emacs or VIM, and not doing any GUI programming. So if the GUI and IDE can be made optional from configure, then the build process may be smoother.

It's a minor issue though. If it can be done easily then I think it's worth doing.

**Comment by dybvig** [comment]
X11 include files are for the expression editor (expeditor), not for SWL, which takes care of its own dependencies (assuming it still works).  The expression editor uses it as a way to support paste (^V) if the corresponding dynamic library is available at run time.  Commenting out line 283 of version.h:

```
#define LIBX11 "/usr/X11R6/lib/libX11.dylib"
```

should disable the includes.  They could be disabled automatically by configure when the include files don't exist, as Yin suggests.

**Comment by yinwang0** [comment]
I just tried the expression editor and it's awesome!

I tried Cmd-C and Cmd-V (not Ctrl-V) on a mac, and it seems to work well without turning on XQuartz server. So I guess the X11 library is only for copy and paste from a X11 program? 

Since it's rare to run a X11 program on Mac, maybe this can indeed be made optional.

**Comment by michaelballantyne** [comment]
It seems like the X11 dependency and the expression editor's support for paste is only relevant if Chez is not running within a terminal emulator that itself supports pasting.

Pasting with CMD-V in iTerm or with the middle mouse button in the X11 terminal on OS X and in Konsole on Linux work without the X11 support. The CTRL-Y action doesn't seem to need X11.

@dybvig Are there environments where having paste support built in to the expeditor rather than provided by the terminal emulator is important?

**Comment by dybvig** [comment]
Alternative paste methods notwithstanding, in order for the X Windows experience to be the same on all machines, using Control-V in an X terminal must paste text selected in the same or another X application, which requires X Windows support to be compiled in.  When we were doing the builds, we made sure the X11 header files were available on the build machine.  What we probably need now is for X Windows support to be disabled via a configure option, which we would ask those who create binary packages not to use (implying they should also have the X11 header files in place).  That would allow people to build their own binaries without X Windows support while those who install packages get the X Windows support if they've also installed XQuartz.  (libX11.dylib is loaded dynamically if present at the first Control-V, so a binary built with X11 support can run with or without XQuartz installed )

**Comment by faywong** [comment]
@yinwang0 When i build it on mac os (configure and make without extra params), this error occurred, hasn't you run across it?
```bash
gcc  -m64 -Wpointer-arith -Wall -Wextra -Werror -O2 -I/opt/X11/include/  -o ../bin/a6osx/scheme ../boot/a6osx/kernel.o ../boot/a6osx/main.o -liconv -lm -lncurses 
Undefined symbols for architecture x86_64:
  "_gzclearerr", referenced from:
      _S_new_open_input_fd in kernel.o
      _S_compress_input_fd in kernel.o
      _S_close_fd in kernel.o
      _S_bytevector_read in kernel.o
      _S_bytevector_write in kernel.o
      _S_put_byte in kernel.o
      _uf_read in kernel.o
      ...
  "_gzclose", referenced from:
      _S_new_open_input_fd in kernel.o
      _S_compress_input_fd in kernel.o
      _S_close_fd in kernel.o
      _find_boot in kernel.o
      _load in kernel.o
  "_gzdirect", referenced from:
      _S_new_open_input_fd in kernel.o
      _S_compress_input_fd in kernel.o
  "_gzdopen", referenced from:
      _S_new_open_input_fd in kernel.o
      _S_compress_input_fd in kernel.o
      _S_compress_output_fd in kernel.o
      _new_open_output_fd_helper in kernel.o
  "_gzerror", referenced from:
      _S_new_open_input_fd in kernel.o
      _S_compress_input_fd in kernel.o
      _S_close_fd in kernel.o
      _S_bytevector_read in kernel.o
      _S_bytevector_write in kernel.o
      _S_put_byte in kernel.o
      _uf_read in kernel.o
      ...
  "_gzgetc", referenced from:
      _find_boot in kernel.o
      _zget_uptr in kernel.o
  "_gzopen", referenced from:
      _find_boot in kernel.o
  "_gzread", referenced from:
      _S_bytevector_read in kernel.o
      _uf_read in kernel.o
      _find_boot in kernel.o
  "_gzrewind", referenced from:
      _find_boot in kernel.o
  "_gzseek", referenced from:
      _S_get_fd_pos in kernel.o
      _S_set_fd_pos in kernel.o
  "_gzungetc", referenced from:
      _find_boot in kernel.o
  "_gzwrite", referenced from:
      _S_bytevector_write in kernel.o
      _S_put_byte in kernel.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by TisonKun** [comment]
With https://github.com/cisco/ChezScheme/commit/ab43608f22dc68497577ab527c782507144ef0e4 and https://github.com/cisco/ChezScheme/commit/b4793db8362152b242d5694417f555efec0db81a, Xlib now is auto-detected while configuring and be a option. It would address this specific issue.

cc @dybvig @yinwang0

**Comment by jltaylor-us** [comment]
Was this issue resolved with #331 ?

**Comment by tisonkun** [comment]
I think so. But I am not one of the maintainer or collaborator who is able to close the issue :P
